python_modules/dagster/dagster/_core/definitions/definitions_class.py

```diff
@@ -730,3 +732,19 @@ def get_all_asset_specs(self) -> Sequence[AssetSpec]:
         """Returns an AssetSpec object for every asset contained inside the Definitions object."""
         asset_graph = self.get_asset_graph()
         return [asset_node.to_asset_spec() for asset_node in asset_graph.asset_nodes]
+
+    @experimental
+    def with_metadata(self, metadata: Mapping[str, Any]) -> Self:
+        """Add metadata to the Definitions object. This is typically used to cache data
+        loaded from some external API that is computed during execution of a DefinitionsLoader in a
+        code server. The cached data is then made available on the DefinitionsLoadContext when the
+        same target is loaded in another context (such as a run worker), allowing use of the cached
+        data to avoid additional external API queries.
+        """
+        return copy(
+            self,
+            metadata={
+                **(self.metadata or {}),
+                **metadata,
+            },
+        )
```

Comment:
<!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: lines +736 to +751]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->shouldn't this be downstack?<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24395?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

Reply:
See my comments in that downstack PR. If we go with "metadata is private attribute only accessed through `with_metadata`", then yes, but I proposed something different.

Reply:
roger